[STN-59] Integrate Decanter 5.x into Sass Build

nrambeck opened the pull request:
## Summary

Added Decanter as a dependency to our Humsci Basic frontend build. The 6.x version is in alpha, so I opted to go with the latest 5.x version. 

Based on feedback from Shea about the direction Decanter is going, I added the `node_modules` directory as an include path instead of including the Decanter `src/scss` directory. This means our imports will reference longer paths, but will make it more obvious where imports are importing from.

To test integration, I imported the `decanter-no-markup` file and used one of the variables from Decanter. I also imported the `brand-bar` component styles.

Additionally, this PR switches our build from using Node Sass to using Dart Sass which has the latest version of Sass which provides some powerful new features. Dart Sass compiles faster as well. The downside is that this new version of Sass is returning deprecation warnings for Sass code in Decanter which will generate a bunch on noise on each build. Something to consider.

## Steps to Test

1. Run `npm install` in the Humsci Basic directory 
1. Run `npm run build:sass` to compile the styles. You will see some deprecation warnings, but the styles should compile as expected. The cardinal red variable from Decanter should compile to the correct color value and the brand bar styles should show up in the compiled stylesheet.
1. Run `npm run watch` to test that the watch setup still works
1. From the root of the project, run `npm run build:css` and `npm run watch` to ensure that Grunt does not try to compile any styles in our Humsci Basic theme.

## See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

Commit:
STN-59 import Decanter variables and mixins, and the brand-bar component styles

Commit:
STN-59 Switch to using Dart Sass instead of Node Sass

Commit:
fix: Exclude humsci_basic theme from Sass compile in root Grunt build

Rename npm script for running the Sass build

nrambeck commented on docroot/themes/humsci/humsci_basic/package.json:
```diff
@@ -14,6 +14,10 @@
     "ie 11"
   ],
+  "dependencies": {
+    "bourbon": "^5.1.0",
```
Bourbon is evidently a required dependency of Decanter.

silverli commented on docroot/themes/humsci/humsci_basic/Gruntfile.js:
```diff
@@ -5,8 +5,9 @@ module.exports = function (grunt) {
-      '**/*.scss',
-      'node_modules/decanter/scss',
+      'src/**/*.scss',
+      'node_modules/bourbon/core',
```
can we write a note about why we have to explicitly add this here? (since decanter must import it like this) and the node_modules line below is apparently not enough context for that import

nrambeck replied:
Great idea. I'll add a comment above this line.

silverli commented on docroot/themes/humsci/humsci_basic/package.json:
```diff
@@ -14,6 +14,10 @@
   ],
+  "dependencies": {
+    "bourbon": "^5.1.0",
+    "decanter": "^5.1.0"
```
@sherakama is decanter 5.1 a good latest version to bring in? Nathan saw that the latest is an alpha release

sherakama replied:
Copying the slack response here:

> As with any major version difference there will be work on your end to adjust from 5 to 6. We are planning to release the first stable version of 6 tomorrow. Nothing too major in version 6 I hope but I haven't put together the full change docs yet. If you are already in flight on your work you are safe to stay on version 5. We will release a final version with all the deprecations for that as well. When you are ready you can then look at the effort to update and plan accordingly.

silverli replied:
Thanks Shea!

silverli commented on docroot/themes/humsci/humsci_basic/package.json:
```diff
@@ -23,10 +23,10 @@
     "autoprefixer": "^9.7.3",
     "grunt": "^1.0.4",
     "grunt-contrib-watch": "^1.1.0",
+    "grunt-dart-sass": "^1.1.3",
```
ha! grunt is staying up to date!! 🎉

silverli commented on docroot/themes/humsci/humsci_basic/package.json:
```diff
@@ -23,10 +23,10 @@
     "grunt-postcss": "^0.9.0",
     "grunt-run": "^0.8.1",
-    "grunt-sass": "^3.1.0",
-    "node-sass": "^4.13.0",
```
👍

silverli commented on Gruntfile.js:
```diff
@@ -5,7 +5,7 @@ module.exports = function (grunt) {
   function getIncludeFiles() {
     var patterns = [
-      'docroot/themes/humsci/**/*.scss',
+      'docroot/themes/humsci/!(humsci_basic)/**/*.scss',
```
looks great

silverli reviewed (changes requested):
This looks great!!

I had one comment I posted in the code inline and just had one issue running the rootfile Grunt task: `npm run watch` - I noticed it is still watching the humsci_basic directory, even if it doesn't compile the sass.

one thought:
the dart-sass deprecation warnings are something we could make into a card to try to create a PR on the decanter side in the new year.

pookmish commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr439-4sfnmnuq6qqx3nu37ejfhurpvufoknbs.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5dfaa1856c12264a260ab117

Visual Diffs:
- https://dashboard.tugboat.qa/5dfaa1856c12264a260ab117/visualdiffs

pookmish commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr439-4sfnmnuq6qqx3nu37ejfhurpvufoknbs.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5dfaa1856c12264a260ab117

Visual Diffs:
- https://dashboard.tugboat.qa/5dfaa1856c12264a260ab117/visualdiffs

silverli commented on Gruntfile.js:
```diff
@@ -27,7 +27,11 @@ module.exports = function (grunt) {
+        files: [
+          '**/*.{scss,sass}',
+          '!**/node_modules/**',
+          '!docroot/themes/humsci/humsci_basic/**',
```
👍

silverli commented on Gruntfile.js:
```diff
@@ -65,7 +69,10 @@ module.exports = function (grunt) {
-            src: ['modules/humsci/**/[a-z]*.scss', 'themes/humsci/**/[a-z]*.scss'],
+            src: [
+              'modules/humsci/**/[a-z]*.scss',
+              'themes/humsci/!(humsci_basic)/**/[a-z]*.scss',
```
👍

silverli commented on docroot/themes/humsci/humsci_basic/Gruntfile.js:
```diff
@@ -5,8 +5,10 @@ module.exports = function (grunt) {
-      '**/*.scss',
-      'node_modules/decanter/scss',
+      'src/**/*.scss',
+      // Decanture uses Bourbon imports relative to this path
```
nice!

silverli commented on docroot/themes/humsci/humsci_basic/Gruntfile.js:
```diff
@@ -24,7 +26,7 @@ module.exports = function (grunt) {
       css: {
         files: ['src/**/*.{scss,sass}'],
-        tasks: ['sass'],
+        tasks: ['dart-sass:dist', 'run:stylelint'], // run stylelint at same time after each change
```
hmm we took this out in the other branch (the stylelint task in the watch)

silverli replied:
as well as the comment

silverli replied:
also the run task for linting is wrong now:

should be lint:sass

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/_base.scss:
```diff
@@ -10,10 +16,12 @@ html {
   }
 }
+
+@import 'decanter/core/src/scss/components/brand-bar/index';
```
do we want to keep this in here? I don't think it hurts but I thought I'd ask since I know this was also brought in just to do some testing of whether it worked

silverli reviewed (changes requested):
This is looking great. And now the root watch task ignores the humsci_basic folder! A couple remaining questions and issues:

- Should we edit the root package.json to match the task names we have in the humsci_basic package.json now? Basically build:sass vs build:css.  I think there should be consistency and that's partly my bad for not realizing before this!
- we also need to update the humsci_basic gruntfile to not have stylelint run during the watch task, as well as update the watch task appropriately to be what's in the package.json now (seems like some merge issues with my branch)
- Also, the sass-tests are failing now because of decanter... hmm we might need to go back to importing the paths the other way?

pookmish commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr439-4sfnmnuq6qqx3nu37ejfhurpvufoknbs.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5dfaa1856c12264a260ab117

Visual Diffs:
- https://dashboard.tugboat.qa/5dfaa1856c12264a260ab117/visualdiffs

nrambeck commented:
Great feedback. 

- I changed the root npm script to `npm run build:sass` to match what we do in our theme
- I fixed the watch task to remove linting (yes that was a merge conflict that got resolved poorly). If there was something else to do on bullet #2, I didn't understand.
- I fixed the tests by adding the same includePaths to the test suite that we have in our Grunt file

silverli commented on docroot/themes/humsci/humsci_basic/Gruntfile.js:
```diff
@@ -24,7 +26,7 @@ module.exports = function (grunt) {
       css: {
         files: ['src/**/*.{scss,sass}'],
-        tasks: ['sass'],
+        tasks: ['dart-sass:dist'],
```
I like this explicit naming

pookmish commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr439-4sfnmnuq6qqx3nu37ejfhurpvufoknbs.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5dfaa1856c12264a260ab117

Visual Diffs:
- https://dashboard.tugboat.qa/5dfaa1856c12264a260ab117/visualdiffs

pookmish commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr439-4sfnmnuq6qqx3nu37ejfhurpvufoknbs.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5dfaa1856c12264a260ab117

Visual Diffs:
- https://dashboard.tugboat.qa/5dfaa1856c12264a260ab117/visualdiffs

silverli commented on docroot/themes/humsci/humsci_basic/Gruntfile.js:
```diff
@@ -24,20 +26,20 @@ module.exports = function (grunt) {
     postcss: {
       options: {
-        map: true, // inline sourcemaps
+        map: false, // inline sourcemaps
```
we decided to remove this for now until we can test it a bit more

silverli reviewed (approved):
Looks great!! Thanks for all the revisions!!
